### PR TITLE
WIP =str JDK7 AsynchronousFileChannel backed FileSource

### PR DESCRIPTION
**Pull request by agolubev**
Ref #17269 
The first try - approach will stay the same, however, will try to add "read ahead" or some other  approaches for optimizations

### CODE REVIEW

**Comment by johanandren** [comment]
@agolubev you expect to add said improvements in this PR? (If so, please add a WIP in front of the PR title for now)

**Comment by jrudolph** [comment]
How does this PR relate to https://github.com/akka/akka/pull/21801? Did you the discussion a while ago on the gitter chat (I think I mentioned you): https://gitter.im/akka/dev/archives/2017/05/18? I think we should first make sure that we know what the goals of this effort is :)

It seems the last time you measured using AsynchronousFileChannel could be slower.

**Comment by agolubev** [comment]
Per @ktoso ticket desription my assumption will be that I need to create performance test for reading simultaneously from 100 files and reading from one file with 100 streams to check how it is scaling

**Comment by agolubev** [comment]
I recall that last time I did #21801 - sync NIO was faster than async. Also async requires separate threads management (that's another drawback).
Anyway - let me know if you have some other test cases to compare in mind.

**Comment by agolubev** [comment]
NIO sync (old) -  FilePublisher - master
```
[info] Benchmark                                (bufSize)  Mode  Cnt    Score    Error  Units
[info] FileSourcesScaleBenchmark.flatMapConcat       2048  avgt    8  808.535 ± 87.077  ms/op
[info] FileSourcesScaleBenchmark.mapAsync            2048  avgt    8  507.474 ± 36.510  ms/op
```
NIO sync (new) - Graph Stage PR#21801
```
[info] Benchmark                                (bufSize)  Mode  Cnt     Score     Error  Units
[info] FileSourcesScaleBenchmark.flatMapConcat       2048  avgt    8  1331.689 ± 353.467  ms/op
[info] FileSourcesScaleBenchmark.mapAsync            2048  avgt    8   517.148 ± 133.549  ms/op
```
NIO async (new) - PR#23217
```
[info] Benchmark                                (bufSize)  Mode  Cnt    Score    Error  Units
[info] FileSourcesScaleBenchmark.flatMapConcat       2048  avgt    8  593.276 ± 61.132  ms/op
[info] FileSourcesScaleBenchmark.mapAsync            2048  avgt    8  375.559 ± 73.783  ms/op
```

**Comment by agolubev** [comment]
Let me know if I need to rebase

**Comment by jrudolph** [comment]
So, what do you suggest, @agolubev? Based on your benchmark there seems to be some improvement to the `flatMapMerge` case but the `mapAsync` case seems to be twice as slow as before. That leaves quite a lot of questions open for me:

 * How do you prevent that the files are read directly from the page cache? Or isn't reading from disk something you want to measure?
 * How does performance change for other levels of parallelism? Especially, how does it change for the single file case?
 * For what OSs do the results hold?
 * What's the reason for the performance difference? (How did you configure the IO dispatcher in the old version?)
 * What kind of effects do you expect because we would be using the JDK-level thread-pool?

If we don't have any good answers to those questions and the benefits are unclear, I'd be reluctant to add extra complexity.

**Comment by agolubev** [comment]
@jrudolph as I'm reading test results it's in ms/op so less is better. Consequently, the only result that is worse is `flatMapMerge` in new sync implementation.

Regarding AsynchronousFileChannel: it uses `Executors.newCachedThreadPool` that creates new threads as needed. The initial size of pool equals to the number of processors. There are two system attributes: `java.nio.channels.DefaultThreadPool.threadFactory` and `java.nio.channels.DefaultThreadPool.initialSize` that allows managing pool from settings  (not to constructing from code). So out of the box it's not a lot of configuration tools. 

Tests now are about reading from 40 different big files in parallel.  For `mapAsync` - it's 40 threads with an individual stream in each. For `flatMapMerge` - it's merging all 40 streams into one (which can be a bottle neck.)

I can partially answer your questions and then list my next steps.
> How do you prevent that the files are read directly from the page cache? Or isn't reading from disk something you want to measure?

Files are fairly big (20mb) so they will not fit into buffer - assuming 32Mb buffer on my mac book pro

>How does performance change for other levels of parallelism? Especially, how does it change for the single file case?

I'm going to add test case for one big file and reading it from the beginning and from several different places

>For what OSs do the results hold?

Tested only on OS X 10.11.6

> What's the reason for the performance difference? (How did you configure the IO dispatcher in the old version?)

This is what I need to try.

> What kind of effects do you expect because we would be using the JDK-level thread-pool?

The test shows that with  AsynchronousFileChannel and JDK level thread pool it's faster even with default settings. So this can be either out of the box thread pool is better than default dispatcher or JDK is doing better job by itself - so this should be proved by the benchmark. 

Basically, we have several dimensions:
1) 3 approaches
2) Reading from many files, reading from one file
3) Checking on different level threads/streams - like 20, 40 to get the trend
4) Play with thread-pool-executor in pool dispatcher to get the best performance for old cases

Things to do:
1) So first of all need to switch to sync new version to see why it's slower than old one.
2) Also need to try dispatcher configuration to make sure it optimized for task
3) Need to create test with multiple reads from big single file

**Comment by jrudolph** [comment]
> as I'm reading test results it's in ms/op so less is better. Consequently, the only result that is worse is flatMapMerge in new sync implementation.

Oh, I see, I read the wrong column.

> Files are fairly big (20mb) so they will not fit into buffer - assuming 32Mb buffer on my mac book pro

Not sure how it is for macos, but on Linux almost the entire memory will be used as cache as long as it is not needed otherwise.

I agree it might make sense to test more dimensions. In the end, we probably cannot optimize for all scenarios but it would be nice to be able to make an informed choice.

**Comment by agolubev** [comment]
>Not sure how it is for macos, but on Linux almost the entire memory will be used as cache as long as it is not needed otherwise.

Make sense. Let me check with iostat to be sure

**Comment by agolubev** [comment]
interim results:
1) flatMapMerge slowness on NIO sync (new) - Graph Stage PR#21801 caused by fusing issue (nio operations were done in the standard dispatcher threads. After rebase to HEAD the issue gone.
2) Increased file size to 60Mb each to see spikes in disk IO activities.   
3) Testing with 40 threads in dispatcher thread pool

**Comment by agolubev** [comment]
Ok here is new statistics.
40 threads (IO dispatcher settings for sync NIO) with 40 Substreams
40 Files 60mb each. 
Per my observation dispersion is pretty big - perhaps because of background activities on my laptop

NIO async (new) - PR#23217
```
[info] Benchmark                                (bufSize)  Mode  Cnt     Score      Error  Units
[info] FileSourcesScaleBenchmark.flatMapConcat       2048  avgt    5  5390.700 ± 1764.132  ms/op
[info] FileSourcesScaleBenchmark.mapAsync            2048  avgt    5  3808.986 ± 3531.616  ms/op
```

NIO sync (new) - Graph Stage PR#21801
```
[info] Benchmark                               (bufSize)  Mode  Cnt     Score      Error  Units
[info] FileSourcesScaleBenchmark.flatMapMerge       2048  avgt    5  5580.241 ± 1065.590  ms/op
[info] FileSourcesScaleBenchmark.mapAsync           2048  avgt    5  5720.858 ± 1178.052  ms/op
```

NIO sync (old) - FilePublisher - master
```
[info] Benchmark                               (bufSize)  Mode  Cnt     Score      Error  Units
[info] FileSourcesScaleBenchmark.flatMapMerge       2048  avgt    5  5947.536 ±  632.910  ms/op
[info] FileSourcesScaleBenchmark.mapAsync           2048  avgt    5  4892.486 ± 3361.482  ms/op
```

for 20 files - it's  for all
flatMapMerge  - 1500 ms/op
mapAsync - 800 ms/op

To Sum up:
Per my understanding async NIO does not provide good performance benefits but brining need to manage java native pool (that does not have a good configuration like in Akka)
I would choose sync NIO as a solution for Akka but would like to hear other opinions.

File: akka-bench-jmh/src/main/scala/akka/stream/io/FileSourcesScaleBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import akka.{ Done, NotUsed }
+import org.openjdk.jmh.annotations.{ BenchmarkMode, OutputTimeUnit, Scope, State, _ }
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future, Promise }
+
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@BenchmarkMode(Array(Mode.AverageTime))
+class FileSourcesScaleBenchmark {
```
Comment (patriknw) [inline akka-bench-jmh/src/main/scala/akka/stream/io/FileSourcesScaleBenchmark.scala]: you could try some of these annotations to run more measurement observations
```
@Warmup(iterations = 10, time = 5, timeUnit = TimeUnit.SECONDS, batchSize = 1)
@Measurement(iterations = 10, time = 15, timeUnit = TimeUnit.SECONDS, batchSize = 1)
```

**Comment by patriknw** [comment]
I think the thread pool argument is good, and the performance is similar or better, so I vote for sync NIO. Meaning that this PR can be closed and we can continue in https://github.com/akka/akka/pull/21801 ?
The benchmark here would be good to include in the other PR.

File: project/MiMa.scala [hunk 2]
```diff
@@ -1229,7 +1229,15 @@ object MiMa extends AutoPlugin {
...
-        ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.remote.transport.AssociationHandle.disassociate")
+        ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.remote.transport.AssociationHandle.disassociate"),
+
+        ProblemFilters.exclude[MissingTypesProblem]("akka.stream.impl.io.FileSource"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.FileSource.newInstance"),
+        ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.stream.impl.io.FileSource.withAttributes"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.FileSource.attributes"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.FileSource.label"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.FileSource.create"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.FileSource.this")
```
Comment (patriknw) [inline project/MiMa.scala]: these should be placed in akka-stream/src/main/mima-filters/2.5.4.backwards.excludes now

**Review by patriknw — APPROVED** [review]
LGTM, I really think we should merge this now. Sorry that we have been so slow in the review process.

mima needs fixing

**Comment by patriknw** [comment]
sorry, I commented on wrong PR, this should be closed in favor of https://github.com/akka/akka/pull/21801
